src/jmath/rand.ts

```diff
@@ -87,9 +70,18 @@ export function chooseObjectWithProbability<T extends objectWithProbability>(
     (maxProbability, current) => current.probability + maxProbability,
     0,
   );
-  // Choose random integer within the sum of all the probabilities
-  const roll = randInt(1, maxProbability, seedRandomInstance);
-  return _chooseObjectWithProbability(roll, source);
+  // Choose random number within the sum of all the probabilities
+  let roll = randFloat(0, maxProbability, seedRandomInstance);
+  // Iterate each object and check if the roll is within the probability range
+  for (let x of source) {
+    if (roll < x.probability) {
+      return x;
+    }
+    roll -= x.probability;
+  }
+
+  console.warn("ChooseObjectWithProbability returned nothing. This probably shouldn't happen?: ", roll, source)
+  return undefined;
```

Comment:
Good catch on finding that randFloat maxInclusive should actually be maxExclusive - I missed that.

That rename change can remain, however, I want you to leave _chooseObjectWithProbability and chooseOneOfSeeded as they were.  This is long-standing, well-trusted engine code (which also has tests already).  I want to avoid engine code changes unless there is something actually wrong.  I'd rather adapt the code that uses the engine code to meet the limitations of the engine code rather than change the engine code to meet every changing need.  I see that you find it useful to have this work with floats so that polymorph can use `probability: (budgetDiff == 0) ? 1 : 1 / budgetDiff`; however it will be much safer to just modify polymorph slightly so that the probability is an integer such as `Math.floor(1000 / budgetDiff)`.
Especially this late in the development of the game, it is a high-risk decision to change engine code like this.  It might be fine, but it might introduce hidden errors and would require a lot of testing and scrutiny.  Opt for the simpler, safer solution instead (unless of course a bug or problem is found in the engine code).